apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt

```diff
@@ -77,74 +74,117 @@ private constructor(
-            )
-          } catch (e: Exception) {
-            // Error opening the websocket
-            mutableEvents.emit(NetworkError(e))
-            continue
-          }
+    override fun generalError(payload: Map<String, Any?>?) {
+      messages.trySend(GeneralError(payload))
+    }
+
+    override fun networkError(cause: Throwable) {
+      messages.trySend(NetworkError(cause))
+    }
+  }
 
-          protocol = protocolFactory.create(
-              webSocketConnection = webSocketConnection,
-              listener = this@WebSocketNetworkTransport,
-          )
-          try {
-            protocol.connectionInit()
-          } catch (e: Exception) {
+  /**
+   * Long-running method that creates/handles the websocket lifecyle
+   */
+  private suspend fun supervise(scope: CoroutineScope) {
+    /**
+     * No need to lock these variables as they are all accessed from the same thread
+     */
+    var idleJob: Job? = null
+    var connectionJob: Job? = null
+    var protocol: WsProtocol? = null
+    var subscriptions = 0
+
+    while (true) {
+      val message = messages.receive()
+
+      when (message) {
+        is Event -> {
+          if (message is NetworkError) {
             protocol = null
-            // Error initializing the connection
-            mutableEvents.emit(NetworkError(e))
-            continue
-          }
-          connectionJob = launch {
-            protocol!!.run(this)
+            connectionJob?.cancel()
+            connectionJob = null
+            idleJob?.cancel()
+            idleJob = null
           }
+          mutableEvents.tryEmit(message)
         }
+        is Command -> {
+          if (protocol == null) {
+            if (message !is StartOperation<*>) {
+              // A stop was received, but we don't have a connection. Ignore it
+              continue
+            }
+
+            val webSocketConnection = try {
+              webSocketEngine.open(
+                  url = serverUrl,
+                  headers = mapOf(
+                      "Sec-WebSocket-Protocol" to protocolFactory.name,
+                  )
+              )
+            } catch (e: Exception) {
+              // Error opening the websocket
+              mutableEvents.emit(NetworkError(e))
+              continue
+            }
 
-        when (command) {
-          is StartOperation<*> -> {
-            subscriptions++
-            protocol.startOperation(command.request)
+            protocol = protocolFactory.create(
+                webSocketConnection = webSocketConnection,
+                listener = listener,
+                scope = scope,
+            )
+            try {
+              protocol.connectionInit()
+            } catch (e: Exception) {
+              // Error initializing the connection
+              protocol = null
+              mutableEvents.emit(NetworkError(e))
+              continue
+            }
+
+            connectionJob = scope.launch(start = CoroutineStart.UNDISPATCHED) {
```

Comment:
Question: why do de need `UNDISPATCHED` here?

Reply:
I wanted to make sure `protocol.run()` would always be called. I'm not 100% sure if `protocol` could be reset before starting to execute the coroutine so added this as an extra precaution. Maybe it's not required but it shouldn't hurt so better safe than sorry...

Reply:
👍 Makes sense, thanks (and TIL 😄)!

Reply:
Side note: Looks like the type inference is not working here. `protocol` should be nullable always. I filed https://youtrack.jetbrains.com/issue/KT-50092